cluster.go

```diff
@@ -1031,6 +1036,66 @@ func (c *Cluster) pin(pin api.Pin, blacklist []peer.ID, prioritylist []peer.ID)
 	if err := isReplicationFactorValid(rplMin, rplMax); err != nil {
 		return false, err
 	}
+	switch pin.Type {
+	case api.DataType:
+		// Fall through
+	case api.ShardType:
+		if !pin.Recursive {
+			return false, errors.New("must pin shards recursively")
+		}
+		if len(pin.Parents) > 1 {
+			return false, errors.New("up to one parent per commitment of a shard")
+		}
+		// In general multiple clusterdags may reference the same shard
+		// and sharder sessions typically update a shard pin's metadata.
+		// Hence we check for an existing shard and carefully update.
+		var empty bool // skip check when state does not yet exist
+		cState, err := c.consensus.State()
+		if err != nil {
+			if strings.Contains(err.Error(), "no state has been agreed upon yet") {
+				empty = true
+			} else {
+				return false, err
+			}
+		}
+		if !empty && cState.Has(pin.Cid) {
+			existing := cState.Get(pin.Cid)
+			// For now all repins of the same shard must use the same
+			// replmax and replmin.  It is unclear what the best UX is here
+			// especially if the same Shard is referenced in multiple
+			// clusterdags.  This simplistic policy avoids complexity and
+			// suits existing needs for shard pins.
+			if existing.ReplicationFactorMin != rplMin ||
```

Comment:
We should take the most conservative approach (that is, the highest value for min and max).

Reply:
Thanks for bringing this up.  I've thought a lot about this and am still somewhat lost.  I thought a lot about taking the most conservative approach but I thought that would look like the highest value for min and the lowest value for max, after all users specify a max value to ensure that cids will not be pinned more than some threshold.  If we take the highest value for max, aren't we violating someone's initial assumption?

If we take the lowest max and the highest min we can potentially get into a situation where our highest min is higher than our lowest max and so we won't be able to satisfy anyone's constraints!   There are plenty of rules we could come up with, take the last, take the first, but I don't know which one is correct and I think that ends up being a UX question.

I kept the rule enforcing that pins stay consistent to enforce a rigid pattern of use that while potentially restrictive would prevent the possibility of confusing situations or implicitly redefining the ideas behind replmin and replmax.  I also did this because enforcing other invariants on repl factors requires more complexity during the unpin process to ensure that repl factors represent the invariant over the *current* set of parents.  I was not sure if this complexity was warranted.

If you'd still like high min and high max or some other invariant I'm happy to do it.  Any other guidance or thoughts is also very appreciated.

Reply:
@hsanjuan could you respond to this comment please?